[CHERI-RISC-V] Update for tag clearing

arichardson opened the pull request:
Catches up with https://github.com/CTSRD-CHERI/sail-cheri-riscv/pull/56

Verified by running TestRIG against sail. While fuzzing against sail I also found some additional divergences that I'll upload fixes for in separate PRs.

arichardson committed:
[cheri-common] Use a CHERI_TAG_CLEAR_ON_INVALID macro

This is cleaner than using TARGET_AARCH64 and will be required to update
CHERI-RISC-V to tag-clearing semantics.

arichardson committed:
[CHERI-RISC-V] Use tag-clearing semantics for most instructions

Changing CHERI_TAG_CLEAR_ON_INVALID to 1 swaps over all functions that
are shared with the Morello model to tag clearing. It does not handled
instructions that are specific to MIPS/RISC-V but that will be added in
the following commits.

arichardson committed:
[CHERI-RISC-V] CFromPtr: use tag-clearing semantics

arichardson committed:
[CHERI-RISC-V] CUnseal: use tag-clearing semantics

arichardson committed:
[CHERI-RISC-V] CBuildCap: use tag-clearing semantics

It is still not quite the same as the sail model when processing sentries,
but the trapping behaviour matches now.

arichardson committed:
[CHERI-RISC-V] CToPtr: use tag-clearing semantics

arichardson committed:
[CHERI-RISC-V] CSeal: use tag-clearing semantics

This should also mostly fix CCSeal.

arichardson committed:
[CHERI-RISC-V] CCopyType: use tag-clearing semantics

The current behaviour does not match sail yet, but this is the first
step towards spec-compliant behaviour.

arichardson committed:
[CHERI-RISC-V] CSealEntry: use tag-clearing semantics

It still doesn't quite match sail due to the old requirement to have
CAP_PERM_EXECUTE, but that is not a useful restriction - non-executable
sentries can be extremely useful as an alternative to otypes.

arichardson committed:
[CHERI-RISC-V] Allow creating non-executable sentries

When I originally added sentries, I only considered the jump use case
and added a trap for non-executable inputs to catch errors early.
However, non-executable sentries can be extremely useful if you use
re-derivation for unsealing instead.

These semantics now match the sail model where the restriction was
removed as part of the tag-clearing changes.

For now this keeps the trapping behaviour for MIPS as that allows running
the existing testsuite before and after this change.

arichardson committed:
[CHERI-RISC-V] CAndPerm: update for tag-clearing semantics

arichardson committed:
[cheri-common] Fix assertion in CSeal with out-of-range otype

The cheri-compressed-cap setters assert that the value is in range, so
we have to mask before updating otype to arbitrary values.

arichardson committed:
[CHERI-RISC-V] CSetFlags: use tag-clearing semantics

arichardson committed:
[cheri-common] CBuildCap: Reject non-canonical encodings

Check that the input capability could have been created using a sequence
of valid subsetting instructions and compare the result to the raw bit
pattern instead of accepting any subset of the authorizing capability.

This matches the sail behaviour.

arichardson committed:
[cheri-common] Make CCSeal match sail when both inputs are untagged

We have to check the second argument first.

arichardson commented:
Should be ready for review now.

arichardson commented:
Includes #219 to ensure the CI can pass.

arichardson commented:
Rebased.

arichardson committed:
Jenkins: Synchronize boot test with CheriBSD logic

Jenkins nested parallel has some rather surprising behaviours - in this
case it seems like a later successful call to junit set the entire build
status back to stable even if there were test failures. Using the same
logic as the CheriBSD job should hopefully make this branch fail as
expected.

qwattash commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -905,35 +976,33 @@ void CHERI_HELPER_IMPL(cfromptr(CPUArchState *env, uint32_t cd, uint32_t cb,
+        result.cr_tag = 0; /* Detag sealed inputs  */
+    }
+    target_ulong new_addr = cbp->cr_base + rt;
+    if (!is_representable_cap_with_addr(cbp, new_addr)) {
```
Do we want to mark unrepresentable things even when we clear the tag? Not sure whether this may result into false-positives when checking for tag loss due to unrepresentable things.

qwattash reviewed (approved):
Looks Good to me.  My only question is whether we should propagate the "became-unrepresentable" information in incoffset when the source cap is invalid. I don't have a specific case for which it is wrong though.

arichardson commented:
> Looks Good to me. My only question is whether we should propagate the "became-unrepresentable" information in incoffset when the source cap is invalid. I don't have a specific case for which it is wrong though.

Ah that is a good point - probably not too useful if it's already untagged.

arichardson committed:
CHERI-RISC-V: Use a bit in xCCSR to report we are tag clearing

This will allow detecting required CHERI features more easily and is
helpful for the transition period where older implementations still exist.

See also: https://github.com/CTSRD-CHERI/sail-cheri-riscv/pull/74

arichardson committed:
Jenkins: Boot the dev branch of CheriBSD for RISC-V

The main branch of CheriBSD has test failures with tag-clearing QEMU as
it expects trapping behaviour and those changes will not be cherry-picked,
so we have to boot dev until main is updated.

arichardson commented:
Test failures are timeouts on the ASAN builds. Merging.
